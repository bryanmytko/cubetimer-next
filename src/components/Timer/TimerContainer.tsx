import { Dispatch, useCallback, useContext, useEffect, useState } from "react";
import useSound from "use-sound";
import { useSession } from "next-auth/react";
import { useMutation, useQuery } from "@apollo/client";
import dynamic from "next/dynamic";

const DynamicTimes = dynamic(() => import("./Times"));
const DynamicClassicModeTimes = dynamic(() => import("./ClassicModeTimes"));
const DynamicClock = dynamic(() => import("./Clock"));
const DynamicClockButton = dynamic(() => import("./ClockButton"));
const DynamicConfirmModal = dynamic(() => import("./ConfirmModal"));
const DynamicPanel = dynamic(() => import("./Panel"));
const DynamicScramble = dynamic(() => import("../Scramble/Scramble"));

import { CREATE_SOLVE_SESSION, SAVE_SOLVE } from "../../graphql/mutations";
import {
  RecordSolveOptions,
  TimerAction,
  TimerActionKind,
  TimerState,
} from "../../types/timer";
import { TimerContext, TimerDispatchContext } from "./TimerContext";
import { useTimerKeyup } from "../../hooks";
import { SETTINGS_FOR_USER } from "../../graphql/queries/settingsForUser";

const AUDIO_DING = "/assets/audio/ding.mp3";
const TICK = 60;

const TimerContainer = () => {
  const [saveSolve, { loading: saveTimeLoading }] = useMutation(SAVE_SOLVE);
  const [createSolveSession, { loading: createSessionLoading }] =
    useMutation(CREATE_SOLVE_SESSION);
  const { data: session, status } = useSession();
  const [playDing] = useSound(AUDIO_DING);
  const timer = useContext(TimerContext) as TimerState;
  const dispatch = useContext(TimerDispatchContext) as Dispatch<TimerAction>;
  const [submitted, setSubmitted] = useState(false);
  const { data, loading } = useQuery(SETTINGS_FOR_USER, {
    skip: !session?.user?.id,
    variables: { userId: session?.user?.id },
  });

  /* This is necessary due to the random nature of the initial scramble.
   * Just trying to set an initial state with a random scramble will
   * cause hydration errors on rerender due to mismatched text
   */
  useEffect(() => {
    dispatch({
      type: TimerActionKind.INITIALIZE,
    });
  }, [dispatch]);

  useEffect(() => {
    if (data?.settingsForUser?.defaultClassicMode)
      dispatch({ type: TimerActionKind.TOGGLE_CLASSIC_MODE });
  }, [data, dispatch]);

  const setButtonLocked = useCallback(
    (value: boolean) => dispatch({ type: TimerActionKind.LOCK, value }),
    [dispatch],
  );

  useEffect(() => {
    timer.classicModeEnabled &&
    timer.solveTimes.length >= timer.classicModeLength
      ? setButtonLocked(true)
      : setButtonLocked(false);
  }, [
    setButtonLocked,
    timer.solveTimes,
    timer.classicModeEnabled,
    timer.classicModeLength,
  ]);

  const toggleConfirmModal = useCallback(() => {
    dispatch({ type: TimerActionKind.TOGGLE_CONFIRM_ACTIVE });
    setButtonLocked(!timer.locked);
  }, [dispatch, setButtonLocked, timer.locked]);

  const getSolveSession = useCallback(
    async ({ userId }: { userId: string }) => {
      if (timer.classicModeEnabled && !timer.solveSessionId) {
        const response = await createSolveSession({
          variables: { userId, size: timer.classicModeLength },
        });
        dispatch({
          type: TimerActionKind.SET_SOLVE_SESSION_ID,
          id: response.data?.createSolveSession.id,
        });
        return response.data?.createSolveSession.id;
      }

      return timer.solveSessionId;
    },
    [
      createSolveSession,
      dispatch,
      timer.classicModeEnabled,
      timer.solveSessionId,
      timer.classicModeLength,
    ],
  );

  const recordSolve = useCallback(
    async (options: RecordSolveOptions = { penalty: 0 }) => {
      const { penalty } = options;
      setButtonLocked(true);
      let solveId;

      if (session) {
        const userId = session.user.id;
        const solveSessionId = await getSolveSession({ userId });
        const response = await saveSolve({
          variables: {
            penalty,
            puzzle: timer.puzzleType,
            scramble: timer.scramble,
            time: timer.time,
            userId,
            solveSessionId,
          },
        });

        solveId = response.data?.createSolve.id;
      }

      dispatch({
        type: TimerActionKind.ADD_TIME,
        penalty,
        solveId,
        scramble: timer.scramble,
      });
      toggleConfirmModal();
      setSubmitted(false);
      setButtonLocked(false);
    },
    [
      dispatch,
      getSolveSession,
      saveSolve,
      session,
      setButtonLocked,
      timer.puzzleType,
      timer.scramble,
      timer.time,
      toggleConfirmModal,
    ],
  );

  const handleKeydown = useCallback(
    (e: KeyboardEvent) => {
      if (e.key === " ") e.preventDefault();
      if (e.key !== " " || timer.locked || submitted) return;

      dispatch({ type: TimerActionKind.READY });
    },
    [dispatch, submitted, timer.locked],
  );

  const handleEscape = useCallback(() => {
    if (timer.confirmActive) {
      dispatch({ type: TimerActionKind.CANCEL_SOLVE });
    }
  }, [dispatch, timer.confirmActive]);

  const handleEnter = useCallback(() => {
    if (submitted) return;
    if (timer.confirmActive) {
      setSubmitted(true);
      recordSolve();
    }
  }, [timer.confirmActive, recordSolve, submitted]);

  const handleSpacebar = useCallback(() => {
    if (timer.locked) return handleEnter();

    timer.inspectionTime && !timer.running
      ? dispatch({ type: TimerActionKind.TOGGLE_INSPECTION })
      : dispatch({ type: TimerActionKind.TOGGLE_RUNNING });

    if (!timer.running) return;

    toggleConfirmModal();
  }, [
    dispatch,
    timer.locked,
    handleEnter,
    timer.inspectionTime,
    timer.running,
    toggleConfirmModal,
  ]);

  const timerKeyup = useTimerKeyup(timer.keyLocked, dispatch, {
    " ": handleSpacebar,
    Escape: handleEscape,
    Enter: handleEnter,
  });

  useEffect(() => {
    window.addEventListener("keydown", handleKeydown);
    window.addEventListener("keyup", timerKeyup);

    return () => {
      window.removeEventListener("keydown", handleKeydown);
      window.removeEventListener("keyup", timerKeyup);
    };
  }, [handleKeydown, timerKeyup]);

  useEffect(() => {
    let interval: NodeJS.Timeout;

    if (timer.inspectionRunning) {
      const countDown = () => {
        if (timer.countdown > 1) {
          dispatch({
            type: TimerActionKind.COUNTDOWN,
            value: --timer.countdown,
          });
        } else {
          clearInterval(interval);
          playDing();
          dispatch({ type: TimerActionKind.TOGGLE_RUNNING });
        }
      };

      interval = setInterval(() => countDown(), 1000);
      return;
    }

    dispatch({
      type: TimerActionKind.COUNTDOWN,
      value: timer.inspectionTime,
    });

    return () => clearInterval(interval);
  }, [
    dispatch,
    setButtonLocked,
    timer.countdown,
    timer.inspectionRunning,
    timer.inspectionTime,
    playDing,
  ]);

  useEffect(() => {
    let interval: NodeJS.Timeout;

    if (timer.running) {
      interval = setInterval(
        () => dispatch({ type: TimerActionKind.TICK_UP }),
        TICK,
      );
    }

    return () => clearInterval(interval);
  }, [dispatch, timer.running]);

  if (loading || status === "loading") return <></>;

  return (
    <>
      <div className={timer.classicModeEnabled ? "col-span-6" : "col-span-5"}>
        <DynamicScramble />
        <DynamicClock />
        <DynamicClockButton handleKeyup={timerKeyup} />
        {timer.classicModeEnabled && <DynamicClassicModeTimes />}
        <DynamicConfirmModal
          action={recordSolve}
          loading={saveTimeLoading || createSessionLoading}
        />
        <DynamicPanel initialClassicModeEnabled={timer.classicModeEnabled} />
      </div>
      {!timer.classicModeEnabled && <DynamicTimes session={session} />}
    </>
  );
};

export default TimerContainer;
