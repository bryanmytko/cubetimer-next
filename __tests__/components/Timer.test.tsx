import { render } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import "@testing-library/jest-dom";

import Scrambler, { mockGenerate } from "../__mocks__/scrambler";
import TimerContainer from "../../src/components/Timer/TimerContainer";

jest.mock("../../lib/scrambler", () => Scrambler);

jest.mock("next-auth/react", () => {
  const originalModule = jest.requireActual("next-auth/react");
  const mockSession = {
    expires: new Date(Date.now() + 2 * 86400).toISOString(),
    user: { username: "admin" },
  };
  return {
    __esModule: true,
    ...originalModule,
    useSession: jest.fn(() => {
      return { data: mockSession, status: "authenticated" };
    }),
  };
});

describe("<TimerContainer>", () => {
  beforeEach(() => {
    Scrambler.mockClear();
    mockGenerate.mockClear();
  });

  it("renders the components", () => {
    const { container } = render(
      <MockedProvider>
        <TimerContainer />
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
