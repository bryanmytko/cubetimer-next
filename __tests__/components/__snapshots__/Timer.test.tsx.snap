// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timer> renders the components 1`] = `
<div>
  <div
    class="col-span-5"
  >
    <div>
      <p
        class="text-white text-2xl text-center mt-6"
      >
        U L2 B2 U B' R' U R' U2 L D2 L' F2 R2 U2 F2 U2 F R' U' L'
      </p>
      <div
        class="w-11/12 py-4 mx-auto mt-6 text-center rounded card bg-neutral-800"
      >
        <span
          class="font-mono font-black text-white timer text-8xl"
          id="timer-screen"
        >
          0:00
        </span>
      </div>
      <button
        class="timer-btn-start block mx-auto mt-6 px-10 py-5 text-3xl 
        rounded-md w-11/12 bg-yellow-300"
        id="timer-btn"
      >
        Press spacebar or click to start!
      </button>
      <div
        class="flex w-11/12 px-4 py-6 mx-auto mt-6 bg-gray-300 rounded card"
      >
        <div
          class="flex-1"
        >
          <p>
            Cubes Solved: 
            0
          </p>
          <p>
            Average: 
            --
          </p>
          <p>
            Ao5: 
            --
          </p>
          <p>
            Ao10: 
            --
          </p>
        </div>
        <div
          class="flex-1 text-right"
        >
          <p>
            Fastest: 
            --
          </p>
          <p>
            Slowest: 
            --
          </p>
          <label
            class="mr-2"
          >
            Cube Size:
          </label>
          <select
            class="mt-1 p-1.5 border-solid border-r-4 border-transparent rounded-md focus:outline-gray-300"
            data-testid="select"
          >
            <option
              value="2x2"
            >
              2x2
            </option>
            <option
              selected=""
              value="3x3"
            >
              3x3
            </option>
            <option
              value="4x4"
            >
              4x4
            </option>
            <option
              value="5x5"
            >
              5x5
            </option>
          </select>
          <div>
            <label
              class="mr-2"
            >
              Inspection Time:
            </label>
            <select
              class="mt-1 p-1.5 border-solid border-r-4 border-transparent rounded-md focus:outline-gray-300"
              data-testid="select"
            >
              <option
                selected=""
                value="0"
              >
                0 sec.
              </option>
              <option
                value="5"
              >
                5 sec.
              </option>
              <option
                value="15"
              >
                15 sec.
              </option>
            </select>
          </div>
          <div>
            <label
              class="mr-2"
              for="classicMode"
            >
              Classic mode:
            </label>
            <input
              id="classicMode"
              type="checkbox"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="w-11/12 mx-auto mt-6 md:mt-12"
  >
    <h3
      class="mx-auto mb-2 text-center text-gray-100 text-md"
    >
      Times
    </h3>
    <div
      class="pt-2 pb-4 overflow-auto rounded bg-neutral-800 snap-y snap-mandatory"
    >
      <ul
        class="w-11/12 pr-4 mx-auto text-gray-300 h-96"
      />
    </div>
  </div>
</div>
`;
